Extend and fix RecHitTools

clelange opened the pull request:
- add bhOffset_,
- fix getLayerWithOffset for new geometry,
- add firstLayerBH(), lastLayerBH()

#### PR description:

This PR introduces some important functionality and fixes in/to the RecHitTools required to eventually remove magic numbers in other places of the HGCal reconstruction code.
Related to https://github.com/cms-sw/cmssw/issues/26225

#### PR validation:

Ran some test code with D41 geometry and got the following output:

```text
bhOffset_: 36
EE layers: 28
FH layers: 22
BH layers: 14
first EE layer: 1
first FH layer: 1
first BH layer: 9
EE layer at z = 10000: 28
FH layer at z = 10000: 24
BH layer at z = 10000: 24
lastLayerEE: 28
lastLayerFH: 50
lastLayerBH: 50
firstLayerBH: 37
```

There are some discrepancies that need to be understood, which, however, seem to be in the geometry, underlining the importance of this PR for geometry validation.
For example: D41 should have 50 HGCal layers only, 22 FH layers, 14 BH layers.
After talking to @bsunanda , the error might also be in my setup and there are no issues in D41 geoemtry. We will investigate, but that's independent of the PR.

clelange committed:
Extend and fix RecHitTools:

- add bhOffset_,
- fix getLayerWithOffset for new geometry,
- add firstLayerBH(), lastLayerBH()

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10257

- This PR adds an extra 16KB to repository

cmsbuild commented:
A new Pull Request was created by  @clelange (Clemens Lange) for master.

It involves the following packages:

RecoLocalCalo/HGCalRecAlgos


@perrotta, @cmsbuild, @kpedro88, @slava77 can you please review it and eventually sign? Thanks.
@edjtscott, @vandreev11, @sethzenz, @felicepantaleo, @rovere, @argiro, @cseez, @pfs, @lgray, @kpedro88 this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
-1

Tested at: c10b31a7135bcdbcf8422cc4a2770193b61b4b8a

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/803/summary.html

I found follow errors while testing this PR

**Failed tests:  RelVals**

* **RelVals**:

When I ran the RelVals I found an error in the following workflows:
27434.0 step3
<pre>runTheMatrix-results/27434.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D35+RecoFullGlobal_2023D35+HARVESTFullGlobal_2023D35/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D35_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D35+RecoFullGlobal_2023D35+HARVESTFullGlobal_2023D35.log

</pre>29034.0 step3
<pre>runTheMatrix-results/29034.0_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D41+RecoFullGlobal_2023D41+HARVESTFullGlobal_2023D41/step3_TTbar_14TeV+TTbar_14TeV_TuneCUETP8M1_2023D41_GenSimHLBeamSpotFull14+DigiFullTrigger_2023D41+RecoFullGlobal_2023D41+HARVESTFullGlobal_2023D41.log

</pre>

cmsbuild commented:
Comparison not run due to runTheMatrix errors (RelVals and Igprof tests were also skipped)

clelange commented:
OK, that's somewhat expected given the layer numbers I gave above. I'll run those workflows myself to figure out why exactly they fail.

clelange committed:
Remove use of cross-package constant and use RecHitTools instead

clelange committed:
Remove use of magic numbers and use RecHitTools instead
- Introduce getEventSetupPerAlgorithm() to HGCalClusteringAlgoBase for
  EventSetup propagation

cmsbuild commented:
The code-checks are being triggered in jenkins.

clelange commented:
After `git cms-checkdeps -a` and adjusting all related packages to use _RecHitTools_ (see the two commits https://github.com/cms-sw/cmssw/pull/27128/commits/bc39efc202c85c70a6698713b62f7fd3a0b8b20e and https://github.com/cms-sw/cmssw/pull/27128/commits/9798a95a94c67e065b43da62aafb38bc5a36489d), I now get a crash due to excessive memory usage, but I'm not sure why this is happening... Will pick this up again next week, but it would help a lot of someone had a look to see if I made an obvious mistake.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10289

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #27128 was updated. @perrotta, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

kpedro88 commented:
hold
(while memory issue is being resolved - I'm going to try to check it today)

cmsbuild commented:
Pull request has been put on hold by  @kpedro88
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

clelange committed:
revert back to previous getLayerWithOffset since hits in HGCalSc contain additional offset already

clelange committed:
Remove use of magic numbers and use RecHitTools directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10291

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #27128 was updated. @perrotta, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

clelange committed:
use trailing underscore for member variable

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10292

- This PR adds an extra 44KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10292/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10292/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly

clelange commented:
The memory issues are fixed now, thanks @kpedro88  for your help!
I locally tested the two workflows that previously failed, and they are fine now.
This PR should be ready for tests and review.

kpedro88 commented:
unhold
@clelange can you apply the suggested code checks?

kpedro88 commented on RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h:
```diff
@@ -65,20 +65,25 @@ enum VerbosityLevel { pDEBUG = 0, pWARNING = 1, pINFO = 2, pERROR = 3 };
 
  //max number of layers
- static const unsigned int maxlayer = 52;
+ unsigned int maxlayer;
```
these member variables should have a trailing underscore

clelange committed:
use trailing underscore in member variables

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10295

- This PR adds an extra 44KB to repository

cmsbuild commented:
Pull request #27128 was updated. @perrotta, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 3bc97f4281e9b6695c5c15960083e6322aa95f9d
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/848/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/848/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 606 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3215458
 - DQMHistoTests: Total failures: 178
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3214946
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

kpedro88 commented:
@clelange are the comparison changes expected/intended?

clelange commented:
@kpedro88 Since I removed a few hardcoded values, which are different e.g. for the D41 geometry, some changes are expected. See e.g. RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h where `lastLayerFH` is now not 40 anymore but 50. Additionally, the thresholds in RecoLocalCalo/HGCalRecProducers/plugins/HGCalImagingAlgo.cc and other places still assume that EE-FH-BH are separated by layer, which is also not true anymore for the new geometry, but only EE is separate from the other two. Therefore, I expect comparison changes, but only for events with activity in the hadronic parts of the calorimeter.

kpedro88 commented:
Well, this gets into the distinction raised in #26225 (which is not currently followed in many places). lastLayerFH should still be 40 if it refers to the fine sampling hadronic section, but it should be 50 if it refers to the silicon part of the hadronic section.

Also, there are now some merge conflicts.

clelange commented:
If I understand the new geometry correctly, this distinction is not by layer, but it's by wafer/sensor. There are fine and coarse sections in each layer of the `HGCalSi` part (what we refer to as _FH_). This means that the code logic needs to be changed such that one does not check the layer, but the wafer/sensor size instead. According to @bsunanda this is currently the only way of checking this---he might add additional functionality at a later point.
Since `HGCalImagingAlgo` and `HGCal3DClustering` will be replaced by `HGCalCLUE` and further TICL developments, I'm not sure it is worth the effort to adjust the code for the former two, and for the latter optimisation studies need to be performed to find new parameters/thresholds, which would require further code changes anyway. What do you think?

I'll rebase onto tonights IB tomorrow to solve the merge conflicts. I'll then also remove the (afaik last remaining) hardcoded values in https://github.com/cms-sw/cmssw/blob/921fbc5bf3a5d3e43f1a89bb621f5b2f03a388ed/RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h#L50-L67

kpedro88 commented:
I think this is part of the confusion. The "front part" of the hadronic section (FH) is mostly silicon, but all fine-grained. The "back part" (BH) is part silicon and part scintillator, but all coarse-grained. The fine/coarse boundary is still fixed at a specific z value; it's just the material boundaries that are now based on radiation contours. We shouldn't refer to the HGCalSi as FH, because it covers parts of both FH and BH. But if the usage here is only meant for HGCalSi, then the numerical value is at least correct.

I agree we shouldn't spend too much effort on outdated algorithms.

Getting rid of the hardcoded values in HGCalRecHitSimpleAlgo.h would be great (but will conflict with #27129...).

bsunanda commented:
Front 8 layers of HE part are all silicon and then one has mixed layers of silicon and scintillators. However the coarse and fine part in a given layer depend on R-value and we have one function in HGCalDDDConstant which gives the type (0, 1, or 2) with inputs u, v and layer number. The function will be available in one of the PR's which is not yet integrated

clelange commented:
OK, so as a simple mitigation I can just implement the BH offset as replacement for `lastLayerFH()` so that the first 8 layers of the hadronic endcap get the _FH_ thresholds and the ones behind the _BH_ (previously all scintillator) thresholds as a starting point for someone to do the optimisation.

Then I'll leave `HGCalRecHitSimpleAlgo.h` untouched for now and we can clean it up once https://github.com/cms-sw/cmssw/pull/27129 is merged.

perrotta commented:
hold
(Besides needing a rebase even now, this PR would further conflict with #27026 in ```RecoLocalCalo/HGCalRecAlgos/src/RecHitTools.cc```. Unless this one is urgent, we will wait for #27026 to go in first.)

cmsbuild commented:
Pull request has been put on hold by  @perrotta
They need to issue an `unhold` command to remove the `hold` state or L1 can `unhold` it for all

clelange committed:
Use firstLayerBH instead of lastLayerFH for clustering thresholds

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10311

- This PR adds an extra 48KB to repository

cmsbuild commented:
Pull request #27128 was updated. @perrotta, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

clelange commented:
@kpedro88 @bsunanda I've resolved the merge conflicts and updated the clustering code to use the first layer of BH as threshold for the clustering.

@perrotta Most changes in #27160 that conflict with this PR are just code formatting, so this is in my opinion very annoying to rebase and fix all the merge conflicts. I would suggest to remove the code formatting changes from #27160 and then run the code-formatting at a later stage or merge this PR first.

perrotta commented:
> @perrotta Most changes in #27160 that conflict with this PR are just code formatting, so this is in my opinion very annoying to rebase and fix all the merge conflicts. I would suggest to remove the code formatting changes from #27160 and then run the code-formatting at a later stage or merge this PR first.

Thank you @clelange for the updates and the rebase.
Clang formatting has to be applied at some point, and it will always interfere with some other ongoing development. You guys are very active in HGCal in these days, and some of these PRs will inevitably conflict with some other one: that's by the way one of the reasons for which we have called for a review of the ongoing works in HGCal reco to be discussed in the next reco meeting.
As of now, we would like to go as quickly as possible towards merging #27160 (which had to be rebased a few times, already), unless there are urgencies for any of the other competitor PRs, in which case we will try to setup a different schedule.

perrotta commented:
@cmsbuild please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 9888313732491ef2b2ff1980722b79df54bfbce5
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/868/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/868/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 661 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3215458
 - DQMHistoTests: Total failures: 86
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3215038
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

cmsbuild commented:
The code-checks are being triggered in jenkins.

clelange commented:
I've run the code formatting in order to make eventual merging with #27160 easier (hopefully).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10327

- This PR adds an extra 84KB to repository

cmsbuild commented:
Pull request #27128 was updated. @perrotta, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 9e3fd338be769e36e46bf0161f4bb64a378a9f81
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/896/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/896/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 657 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3237932
 - DQMHistoTests: Total failures: 85
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3237513
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

perrotta commented:
unhold

kpedro88 commented:
+upgrade

perrotta commented:
@clelange 

Having updated the clang format here makes quite hard to review the code and disentangle the "real" updates made in this PR.

Further, this PR will conflict with #27049, which is basically ready to go (and once merged will simplify also the modifications needed by this PR, as there will be a few files that are now modified here and which will simply get removed).

My suggestion is:
- Wait for #27049 being merged, as well as #27129
- Rebase this one, and if #27129 has been already merged you can also update HGCalRecHitSimpleAlgo.h here
- Then there will be a race between this PR and  #27160, and the second one being merged will likely have to rebase once again (maybe not, since you already applied clang-format here... this will have to be verified).

clelange commented:
@perrotta 

Fine with me to wait for #27049 and #27129 - I'm watching both of them. If it helps simplify the review, I can also just undo the clang format by moving one commit back, you review the changes, and then I apply clang format again before the merge (and after the tests). This would eventually help #27160 getting merged. Let me know.

perrotta commented:
Clemens Lange <notifications@github.com> ha scritto:

> @perrotta
>
> Fine with me to wait for #27049 and #27129 - I'm watching both of  
> them. If it helps simplify the review, I can also just undo the  
> clang format by moving one commit back, you review the changes, and  
> then I apply clang format again before the merge (and after the  
> tests). This would eventually help #27160 getting merged. Let me know.

Indeed that would help...
Thank you,
Andrea

> --
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub:
> https://github.com/cms-sw/cmssw/pull/27128#issuecomment-502653965

clelange commented:
OK, no problem, I've reverted the last commit, and can then re-apply clang format after the code review.

perrotta commented:
@clelange : I'm fine with the code review now, and also with the comparison results, as all differences were already addressed by you and @kpedro88 earlier on in this thread. Please go ahead with the clang formatting, which I think it is nowadays necessary to restart the tests.

Well, the natural order of things would be to merge #27129 first and then also update HGCalRecHitSimpleAlgo.h here. @fabiocos please let us know which are the plans for #27129, which is already fully signed by now: please notice that there is also #27160 in the queue, waiting for this PR being merged.

clelange committed:
Apply code-format

clelange commented:
OK, I'll get to it, just rebasing onto the latest IB and then looking into `HGCalRecHitSimpleAlgo.h`. I'll let you know once done.

clelange committed:
Move to using RecHitTools including code-format

cmsbuild commented:
The code-checks are being triggered in jenkins.

fabiocos commented:
@perrotta @clelange #27129 has been integrated, so now you may update this PR accordingly

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10423

- This PR adds an extra 200KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-checks**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10423/code-checks.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10423/code-checks.patch | patch -p1`
You can also run `scram build code-checks` to apply code checks directly
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10423/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10423/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

clelange committed:
Run code-format and merge with nose changes by Sunanda

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10425

- This PR adds an extra 108KB to repository

Code check has found code style and quality issues which could be resolved by applying following patch(s)
- **code-format**:
https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10425/code-format.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10425/code-format.patch | patch -p1`
You can also run `scram build code-format` to apply code format directly

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10426

- This PR adds an extra 212KB to repository

cmsbuild commented:
Pull request #27128 was updated. @perrotta, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

clelange commented:
@perrotta @kpedro88 I've rebased my changes on top of `official-cmssw/master` to merge with the nose-detector changes by @bsunanda. The `HGCalRecHitSimpleAlgo.h` now also uses the RecHitTools inheriting from https://github.com/cms-sw/cmssw/blob/8f2c1e4dfe8695aca4d3d238114b00e4d6405917/RecoLocalCalo/HGCalRecProducers/interface/HGCalClusteringAlgoBase.h#L86 (I've added it there).
The only non-formatting changes in `HGCalClusteringAlgoBase.h`, `HGCalRecHitSimpleAlgo.h` and `HGCalRecHitWorkerSimple.cc` are related to the use of RecHitTools, see the last two commits.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 8f2c1e4dfe8695aca4d3d238114b00e4d6405917
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/1015/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/1015/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5242 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3254096
 - DQMHistoTests: Total failures: 1001
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3252761
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

perrotta commented:
The tests show now a much larger disagreement with the baseline than the previous version of this PR, in some case even quite significant. See for example in wf 27434.0:

![image](https://user-images.githubusercontent.com/4069749/59740868-a640d980-9269-11e9-9607-5729c189b717.png)

which was

![image](https://user-images.githubusercontent.com/4069749/59740900-bce73080-9269-11e9-8ecf-8caddfaf6ea3.png)

What is changed is the baseline: therefore, I suspect that you are reverting some change that was applied in the already merged pull requests.

@clelange , could you please check?

kpedro88 commented:
@perrotta I rather suspect that rebasing to master means that the changes from another PR were included on top of this one, let's test again

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

kpedro88 commented:
@clelange okay, in this case I would try to make sure the exact previous behavior is preserved

cmsbuild commented:
+1
Tested at: 8f2c1e4dfe8695aca4d3d238114b00e4d6405917
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/1016/summary.html

cmsbuild commented:
Comparison job queued.

perrotta commented on RecoLocalCalo/HGCalRecAlgos/interface/HGCalRecHitSimpleAlgo.h:
```diff
@@ -19,70 +19,48 @@
-    float energy = (hfnose ? (uncalibRH.amplitude() * weightsNose_[layer] * 0.001f) :
-                    (uncalibRH.amplitude() * weights_[layer] * 0.001f));
-    float time   = uncalibRH.jitter();
+    if (DetId::Forward == baseid.det() && HFNose == baseid.subdetId())
```
```C++
    bool hfnose = DetId::Forward == baseid.det() && HFNose == baseid.subdetId();
```

clelange replied:
Done

perrotta reviewed (commented):
Only in case you are going to modify anything else ....

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/1016/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5244 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3254096
 - DQMHistoTests: Total failures: 991
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3252771
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

clelange committed:
Finalise move to RecHitTools for HGCalRecHit algorithms

cmsbuild commented:
The code-checks are being triggered in jenkins.

clelange commented:
@perrotta @kpedro88 With some help from @rovere I fixed the use of RecHitTools in the RecHit calibration algorithms (they were not initialised due to the somewhat weird inheritence structure) so that they output is now equivalent to before. I tested this for one geometry, the central tests should hopefully confirm this for the others as well.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-27128/10506

- This PR adds an extra 120KB to repository

cmsbuild commented:
Pull request #27128 was updated. @perrotta, @cmsbuild, @kpedro88, @slava77 can you please check and sign again.

kpedro88 commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 68b182e955207f2114f5f6ef8c97516007dfce8a
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/1084/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-ce1698/1084/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 659 differences found in the comparisons
 - DQMHistoTests: Total files compared: 33
 - DQMHistoTests: Total histograms compared: 3254096
 - DQMHistoTests: Total failures: 73
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3253689
 - DQMHistoTests: Total skipped: 334
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 32 files compared)
 - Checked 137 log files, 14 edm output root files, 33 DQM output files

perrotta commented:
+1
- It allows to centrally retrieve HGCal layer bounds, and fixes offsets for them
- The changes in the output of jenkis tests show the effect of those fixes in the affected Phase2 geometries, as confirmed in the github thread starting from https://github.com/cms-sw/cmssw/pull/27128#issuecomment-500437990
- Further differences that showed us during the developing phase of this PR were found to be spurious and fixed with the last commit https://github.com/cms-sw/cmssw/pull/27128/commits/68b182e955207f2114f5f6ef8c97516007dfce8a
   - Jenkins tests do not show unexpected differences wrt the baseline any more
- This PR will likely conflict with #27160, which is also supposed to get ready soon

perrotta commented:
@kpedro88 please have a look

kpedro88 commented:
+upgrade

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

rovere commented:
Now that all the mandatory signatures are there, it would be nice integrating this asap, so that #27160 could be rebased for the last time, hopefully, and quickly reviewed and integrated itself.
Thanks.

fabiocos commented:
@rovere indeed there will be a conflict between this PR and #27160, independent on the relative order of integration. At this point I will integrate it in next IB

rovere commented:
@fabiocos Yes, I was suggesting to put this in soon, so that the bigger #27160 could be fixed and reviewed. Thanks for the support in this 'messy' period.

fabiocos commented:
+1
